src/graph/GoExecutor.cpp

```diff
@@ -535,44 +548,42 @@ void GoExecutor::maybeFinishExecution(RpcResponse &&rpcResp) {
             continue;
         }
         auto *eschema = resp.get_edge_schema();
-        if (eschema == nullptr) {
-            continue;
-        }
         std::unordered_map<EdgeType, std::shared_ptr<ResultSchemaProvider>> schemas;
-        std::transform(eschema->cbegin(), eschema->cend(), std::inserter(schemas, schemas.begin()),
-                [] (auto &s) {
-            return std::make_pair(s.first, std::make_shared<ResultSchemaProvider>(s.second));
-        });
-
-        for (auto &vdata : *vertices) {
-            for (auto &edge : vdata.edge_data) {
-                RowSetReader rsReader(schemas[edge.type], edge.data);
-                auto iter = rsReader.begin();
-                while (iter) {
-                    VertexID dst;
-                    EdgeRanking rank;
-                    auto rc = iter->getVid(_DST, dst);
-                    if (rc != ResultType::SUCCEEDED) {
-                        doError(Status::Error("Get dst error when go reversely."));
-                        return;
-                    }
-                    rc = iter->getVid(_RANK, rank);
-                    if (rc != ResultType::SUCCEEDED) {
-                        doError(Status::Error("Get rank error when go reversely."));
-                        return;
-                    }
-                    auto type = std::abs(edge.type);
-                    auto &edgeKeys = edgeKeysMapping[type];
-                    edgeKeys.emplace_back();
-                    edgeKeys.back().set_src(dst);
-                    edgeKeys.back().set_dst(vdata.get_vertex_id());
-                    edgeKeys.back().set_ranking(rank);
-                    edgeKeys.back().set_edge_type(type);
-                    ++iter;
-                }
-            }
+        if (eschema != nullptr) {
+            std::transform(eschema->cbegin(),
+                           eschema->cend(),
+                           std::inserter(schemas, schemas.begin()), [] (auto &s) {
+                return std::make_pair(s.first, std::make_shared<ResultSchemaProvider>(s.second));
+            });
         }
-    }
+        for (auto& vdata : *vertices) {
+            for (auto& edata : vdata.edge_data) {
+                for (auto& edge : edata.get_edges()) {
+                    auto dst = edge.get_dst();
+                    std::unique_ptr<RowReader> reader;
+                    if (!schemas.empty()) {
+                        auto it = schemas.find(edata.type);
+                        DCHECK(it != schemas.end());
+                        reader = RowReader::getRowReader(edge.props, it->second);
```

Comment:
the reader just use once in the loop, can we reload the RowReader::getRowReader and just return a static reader, so it will not malloc and free reader in every loop.

Reply:
Good point.